### DISCUSSION TITLE
some questions or new features

### DISCUSSION THREAD

**Discussion by davidxiaobj** [rodrigo-brito/ninjabot#130] [Q&A] [answered]
Here are some questions or new features

- how to implement trailing stop tool? it'd be great if can share more info?

- what's thoughts for polling orders from exchange every second? how about using user data stream?

- in strategy, currently Timeframe is a single frame, how about i will use both `15m` and `1h` for one asset for example? and if so, maybe need to resample the candles?

- how about adding hooks in strategy, for scenarios: when putting order, i'd like to refer to outside sources, etc.

- any thoughts about integrating with 3rd frontend tool? like tradingview, or open source tools, etc?

**Comment by rodrigo-brito** [comment, marked as answer]
Hi @davidxiao

> how to implement trailing stop tool? it'd be great if can share more info?

I will include a new feature soon, strategies will have access to partial candles, with this feature will be easy to create this tool.

> what's thoughts for polling orders from exchange every second? how about using user data stream?

The pooling is very optimized, we fetch data only when the bot have `NEW`, `PENDING`, and `PARTIAL FILLED` orders in the storage. But we can include this feature, but it is not a priority for now.

> in strategy, currently Timeframe is a single frame, how about i will use both 15m and 1h for one asset for example? and if so, maybe need to resample the candles?

Its very complex to provide multiple timeframes for a given strategy, you can use a small timeframe and resample it in the `Indicator` function to create new samples. I wonder to keep the strategy interface simplest as possible.

> how about adding hooks in strategy, for scenarios: when putting order, i'd like to refer to outside sources, etc.

You can use OrderFeeds and subscribe it in the ninjabot `Run()`, with the option, `ninjabot.WithOrderSubscription`. Any struct that implements the bellow interface is valid to receive orders updates. It's the same for candles (`WithCandleSubscription`). Chart module uses this approach.

```go
type OrderSubscriber interface {
	OnOrder(model.Order)
}
```

> any thoughts about integrating with 3rd frontend tool? like tradingview, or open source tools, etc?

I will try to create an isolated frontend with more features, such as custom indicatos, real-time charts, but it is in the roadmap. You can subscribe for update here: https://github.com/rodrigo-brito/ninjabot/issues/124

**Comment by davidxiaobj** [comment]
awesome, mate, 

Yeah, agreed

thanks very much,

pls let me know if there is anything I can do,

**Reply by rodrigo-brito** [reply]
>  how about using user data stream?

Can you work on this feature?

**Reply by davidxiaobj** [reply]
yeah, sure, any tool used to discuss about how to do it before taking action? slack, etc?

**Reply by rodrigo-brito** [reply]
We can use this issue to discuss some ideas: https://github.com/rodrigo-brito/ninjabot/issues/131
